[css-anchor-position-1] Support ::before/::after as anchors

anttijk opened the pull request:
#### f16c267dbdee11b9090e666477a9fd5dd57756f6
<pre>
[css-anchor-position-1] Support ::before/::after as anchors
<a href="https://bugs.webkit.org/show_bug.cgi?id=281509">https://bugs.webkit.org/show_bug.cgi?id=281509</a>
<a href="https://rdar.apple.com/137975699">rdar://137975699</a>

Reviewed by Alan Baradlay.

Also simplify the data structures and logic.

* LayoutTests/TestExpectations:

Scrolling feature is not implemented yet, passes were not real.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/pseudo-element-anchor-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/pseudo-element-anchor-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::willBeDestroyed):
(WebCore::RenderBox::styleWillChange):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::styleWillChange):

Refactor to track anchors as RenderBoxModelObjects instead of Elements.
Register and unregister anchors.

* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::registerAnchor):
(WebCore::RenderView::unregisterAnchor):
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::evaluate):

Remove FinishedCollectingAnchorNames state which is not needed.
Add name from position-anchor property here too.

(WebCore::Style::penultimateContainingBlockChainElement):
(WebCore::Style::isAcceptableAnchorElement):
(WebCore::Style::findLastAcceptableAnchorWithName):
(WebCore::Style::collectAnchorsForAnchorName):

Build a temporary map of anchor names using currently registered anchors.

(WebCore::Style::AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):

Move the state update code from TreeResolver here.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleScope.cpp:

Remove anchor maps. AnchorPositioningState is the only persistent state now.
This will make invalidation etc easier.

(WebCore::Style::Scope::clearAnchorPositioningState):
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolve):

If we have performanced an interleaved layout update the anchor states.

(WebCore::Style::TreeResolver::updateAnchorPositioningState):

Almost nothing needs to be done here now.
We also avoid one unnecessary resolution round.

Canonical link: <a href="https://commits.webkit.org/285254@main">https://commits.webkit.org/285254@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69aa3dd9f773b752122dc8fb0ad9819c3620ca7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72027 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24812 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76187 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23236 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74142 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59248 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23056 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15326 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75094 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46646 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62046 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/37262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43309 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19519 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21582 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65211 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19880 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77869 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16267 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19052 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16313 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62069 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15912 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12749 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6395 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47245 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48314 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49601 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48058 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
https://searchfox.org/wubkat/rev/cde6cf2e21001cb4055a5f891ce3d7abe5eee3e9/Source/WebCore/style/AnchorPositionEvaluator.cpp#388-390

Is this still up-to-date?

tuankiet65 commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -510,37 +502,87 @@ static bool isAcceptableAnchorElement(Ref<const Element> anchorElement, Ref<cons
+    std::optional<AnchorsForAnchorName> anchorsForAnchorName;
+    for (auto elementAndState : document.styleScope().anchorPositionedStates()) {
+        if (!anchorsForAnchorName)
+            anchorsForAnchorName = collectAnchorsForAnchorName(document);
```
We can just collect anchors before the loop right? Or this is to avoid doing any work if there's no anchor positioned elements?

anttijk replied:
Yeah, this avoid collection in the (common) no-anchors case

anttijk replied:
Changed it to simply test if the map is empty first.

anttijk committed:
[css-anchor-position-1] Support ::before/::after as anchors
https://bugs.webkit.org/show_bug.cgi?id=281509
rdar://137975699

Reviewed by Alan Baradlay.

Also simplify the data structures and logic.

* LayoutTests/TestExpectations:

Scrolling feature is not implemented yet, passes were not real.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/pseudo-element-anchor-dynamic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/pseudo-element-anchor-expected.txt:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::willBeDestroyed):
(WebCore::RenderBox::styleWillChange):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::styleWillChange):

Refactor to track anchors as RenderBoxModelObjects instead of Elements.
Register and unregister anchors.

* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::RenderView::registerAnchor):
(WebCore::RenderView::unregisterAnchor):
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::evaluate):

Remove FinishedCollectingAnchorNames state which is not needed.
Add name from position-anchor property here too.

(WebCore::Style::penultimateContainingBlockChainElement):
(WebCore::Style::isAcceptableAnchorElement):
(WebCore::Style::findLastAcceptableAnchorWithName):
(WebCore::Style::collectAnchorsForAnchorName):

Build a temporary map of anchor names using currently registered anchors.

(WebCore::Style::AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement):
(WebCore::Style::AnchorPositionEvaluator::updateAnchorPositioningStatesAfterInterleavedLayout):

Move the state update code from TreeResolver here.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleScope.cpp:

Remove anchor maps. AnchorPositioningState is the only persistent state now.
This will make invalidation etc easier.

(WebCore::Style::Scope::clearAnchorPositioningState):
* Source/WebCore/style/StyleScope.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolve):

If we have performanced an interleaved layout update the anchor states.

(WebCore::Style::TreeResolver::updateAnchorPositioningState):

Almost nothing needs to be done here now.
We also avoid one unnecessary resolution round.

Canonical link: https://commits.webkit.org/285254@main
